### PR TITLE
feat: Implement --use-index & SPA flags

### PR DESCRIPTION
**Pull request by Antosser**
Implemented the --use-index flag. If the path matches a directory and an index.html is found inside, the index.html gets sent instead of the explorer of the directory.

Implemented the --spa flag. If a file/directory is not found, the index.html in the root gets sent. The existence of index.html in the root is also checked at startup. An error is thrown otherwise:
```rust
        if config.spa() {
            let mut index_html = config.root_dir();
            index_html.push("index.html");

            if !index_html.exists() {
                eprintln!(
                    "{}",
                    "SPA flag is enabled, but index.html in root does not exist. Quitting...".red()
                );
                exit(1);
            }
        }
```

The docs were updated to match the new changes

### CODE REVIEW

**Comment by LeoBorai** [comment]
@Antosser can we have 2 PRs for this? And tests as well please?

Its easier to review and to ensure everything is working as expected.

**Comment by Antosser** [comment]
> @Antosser can we have 2 PRs for this? And tests as well please?
> 
> Its easier to review and to ensure everything is working as expected.

To split this up into 2 PRs, the second will need to come after the first one. They cannot coexist, because of potential merge conflicts. It would be faster to leave this PR as is. I can write test though.

**Comment by Antosser** [comment]
Why do the integration tests not get ran?

Edit: And why do 7 of them fail

**Comment by Antosser** [comment]
An integration test is the right type of test to test the `--spa` and `--use-index` flag. However, I do not understand why they do not get run if you do `cargo test`. Also, given from what I've read, the `#[cfg(test)]`, the `mod ... {`, and the `mod.rs` file inside `tests`, seem unnecessary, see [the docs](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests)

**Comment by Antosser** [comment]
I still need an answer

**Comment by LeoBorai** [comment]
> An integration test is the right type of test to test the `--spa` and `--use-index` flag. However, I do not understand why they do not get run if you do `cargo test`. Also, given from what I've read, the `#[cfg(test)]`, the `mod ... {`, and the `mod.rs` file inside `tests`, seem unnecessary, see [the docs](https://doc.rust-lang.org/book/ch11-03-test-organization.html#integration-tests)

Hey @Antosser!

I recommend reading the workflow files to understand how every test is executed.

With regards to the organization of tests, I suggest leaving them as they are and discuss about them separately.

Lately, for E2E testing I would use something like Bats. I could set in up for you and leave a couple examples for you to follow and apply on this PR.

**Comment by Antosser** [comment]
> I recommend reading the workflow files to understand how every test is executed.

Do you mean the GitHub workflow files or the scripts?

> Lately, for E2E testing I would use something like Bats

Bats, as in .bat files? Couldn't finds anything else with this name

**Comment by Antosser** [comment]
@EstebanBorai 

I want to say that the reason people use this app is that it is the rust alternative to NPM's http-server. People expect, that it is blazingly fast, very reliable, easy to use, and feature-rich. It has to represent the Rust community very well. That's why I want to make this app as friendly and easy to use as possible.

The purpose of NPM's http-server is to quickly set up an HTTP server to host HTML files, and that's the default. In my opinion, we should make that the default too and replace the `--index` flag with an `--explorer` flag.

I really want to add new features, but have no interest in writing non-rust bats tests. Can you please write them, so I can move on to other features without worrying about merge conflicts? Also, please consider using `index.html`s by default.

Thanks for the overall help!

**Comment by LeoBorai** [comment]
Hey @Antosser can you rebase please? I would like to merge this once approved!

**Comment by Antosser** [comment]
I love git

**Comment by LeoBorai** [comment]
> I love git

LOL

**Comment by LeoBorai** [comment]
@Antosser can we have workflow changes in a separato PR please?

File: .github/workflows/build.yml
Comment (LeoBorai) [inline .github/workflows/build.yml:1]: Move to a scoped PR please!

File: .github/workflows/clippy.yml
Comment (LeoBorai) [inline .github/workflows/clippy.yml:1]: Move to a scoped PR please!

File: .github/workflows/fmt.yml
Comment (LeoBorai) [inline .github/workflows/fmt.yml:1]: Move to a scoped PR please!

File: .github/workflows/test.yml
Comment (LeoBorai) [inline .github/workflows/test.yml:1]: Move to a scoped PR please!

File: LICENSE-MIT
Comment (LeoBorai) [inline LICENSE-MIT:1]: Can we have a specific PR for these updates? Im trying to narrow down this PR scope.

File: README.md
Comment (LeoBorai) [inline README.md:1]: I think its okay to add these changes as part of the LICENSE PR!

File: Cargo.toml [hunk 1]
```diff
@@ -58,6 +58,7 @@ tokio = { version = "1.29.1", features = [
 tokio-rustls = "0.23.4"
 toml = "0.7.6"
 humansize = "2.1.3"
+colored = "2.0.4"
```
Comment (LeoBorai) [inline Cargo.toml]: Can we add colors in another PR? So we revisit all places and keep UI consistent?

**Review by LeoBorai — COMMENTED** [review]
Lets move changes not related to SPA/index to a separate PR please!

**Comment by Antosser** [comment]
I never changed any of the workflow files, but it still looks like I changed them in this PR

File: README.md
Comment (Antosser) [reply README.md:1]: Don't remember ever changing the year

File: LICENSE-MIT
Comment (Antosser) [reply LICENSE-MIT:1]: I don't remember ever changing the license.

**Comment by LeoBorai** [comment]
> I never changed any of the workflow files, but it still looks like I changed them in this PR

I think it has a couple issues while rebasing!

**Comment by LeoBorai** [comment]
Perhaphs you can delete those workflow files from your PR?

**Comment by Antosser** [comment]
> Perhaphs you can delete those workflow files from your PR?

If I delete them, and you merge the PR, they will also disappear in the main branch.  
At this point, I have no idea how to fix this PR. I'll just copy/paste the changes I actually made into a new PR

**Comment by Antosser** [comment]
@EstebanBorai Can I make the `--index` flag the default and replace it with a `--explorer` flag? Better to know now before making changes I would have to change again later

**Comment by LeoBorai** [comment]
> @EstebanBorai Can I make the `--index` flag the default and replace it with a `--explorer` flag? Better to know now before making changes I would have to change again later

I would stick to current approach for now. I dont want to introduce unexpected changes for `v0`.

I have plans to move into Axum for a `v1`. I think this change of behavior will fit just fine when on Axum!

**Comment by LeoBorai** [comment]
> > Perhaphs you can delete those workflow files from your PR?
> 
> If I delete them, and you merge the PR, they will also disappear in the main branch. At this point, I have no idea how to fix this PR. I'll just copy/paste the changes I actually made into a new PR

SGTM

**Comment by Antosser** [comment]
> > @EstebanBorai Can I make the `--index` flag the default and replace it with a `--explorer` flag? Better to know now before making changes I would have to change again later
> 
> I would stick to current approach for now. I dont want to introduce unexpected changes for `v0`.
> 
> I have plans to move into Axum for a `v1`. I think this change of behavior will fit just fine when on Axum!

aight

**Pull request closed without merging**